packages/angular_devkit/build_angular/src/browser/tests/behavior/rebuild-errors_spec.ts

```diff
@@ -5,13 +5,267 @@
+                expect(logs).not.toContain(
+                  jasmine.objectContaining<logging.LogEntry>({
+                    message: jasmine.stringMatching(typeErrorText),
+                  }),
+                );
+
+                // Make an unrelated change to verify error cache was updated
+                // Should continue showing no error
+                await harness.modifyFile('src/main.ts', (content) => content + '\n');
+
+                break;
+              case 4:
+                expect(result?.success).toBeTrue();
+                expect(logs).not.toContain(
+                  jasmine.objectContaining<logging.LogEntry>({
+                    message: jasmine.stringMatching(typeErrorText),
+                  }),
+                );
+
+                break;
+            }
+          }),
+          take(5),
+          count(),
+        )
+        .toPromise();
+
+      expect(buildCount).toBe(5);
+    });
+
+    it('detects template errors with AOT codegen differences', async () => {
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        aot: true,
+        watch: true,
+      });
+
+      const typeErrorText = `Type 'string' is not assignable to type 'number'.`;
+
+      // Create a directive and add to application
+      await harness.writeFile('src/app/dir.ts', `
+        import { Directive, Input } from '@angular/core';
+        @Directive({ selector: 'dir' })
+        export class Dir {
+          @Input() foo: number;
+        }
+      `);
+      await harness.writeFile('src/app/app.module.ts', `
+        import { NgModule } from '@angular/core';
+        import { BrowserModule } from '@angular/platform-browser';
+        import { AppComponent } from './app.component';
+        import { Dir } from './dir';
+        @NgModule({
+          declarations: [
+            AppComponent,
+            Dir,
+          ],
+          imports: [
+            BrowserModule
+          ],
+          providers: [],
+          bootstrap: [AppComponent]
+        })
+        export class AppModule { }
+      `);
+
+      // Create app component that uses the directive
+      const goodComponentContent = `
+        import { Component } from '@angular/core'
+        @Component({
+          selector: 'app-root',
+          template: '<dir [foo]="123">',
+        })
+        export class AppComponent { }
+      `;
+      await harness.writeFile('src/app/app.component.ts', goodComponentContent);
+
+      const buildCount = await harness
+        .execute({ outputLogsOnFailure: false })
+        .pipe(
+          timeout(60000),
+          concatMap(async ({ result, logs }, index) => {
+            switch (index) {
+              case 0:
+                expect(result?.success).toBeTrue();
+
+                // Update directive to use a different input type for 'foo' (number -> string)
+                // Should cause a template error
+                await harness.writeFile('src/app/app.component.ts', `
```

Comment:
Changing the component itself directly affects its TTC file, so this is a trivial case.

A more challenging case would be to have an additional directive that also matches the `dir` element, which initially has a `foo: string;` field which becomes an `@Input` in an incremental update. That does not affect its TS shape (I believe), but it implicitly affects `AppComponent` which now writes the `foo` binding into both directives; one of which is incorrect. This scenario should be covered by this logic:

```ts
          // If required to emit, diagnostics may have also changed
          if (!ignoreForDiagnostics.has(sourceFile)) {
            affectedFiles.add(sourceFile);
          }
```

and I'm not sure if the current tests would fail without this logic, whereas I'd expect the more complex scenario to fail without that logic.